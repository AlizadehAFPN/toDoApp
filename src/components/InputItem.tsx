import {
  View,
  Text,
  TouchableOpacity,
  TextInput,
  StyleSheet,
  GestureResponderEvent,
} from 'react-native';

interface itemType {
  onBTNPressed: (event: GestureResponderEvent) => void;
  onChangeText: Function;
  placeHolder: string;
  value: string;
  btnName: string;
}

const InputItem = ({
  onChangeText,
  onBTNPressed,
  placeHolder,
  value,
  btnName,
}: itemType) => {
  return (
    <View style={styles.mainView}>
      <TextInput
        testID='myinput'
        placeholderTextColor={'#5C5C5C'}
        value={value}
        onChangeText={txt => onChangeText(txt)}
        placeholder={value?.length != 0 ? '' : placeHolder}
        style={[styles.input, {paddingLeft: 16}]}
      />
      <TouchableOpacity onPress={onBTNPressed} style={styles.btn}>
        <Text style={styles.addTxt}>{btnName}</Text>
      </TouchableOpacity>
    </View>
  );
};

export default InputItem;

const styles = StyleSheet.create({
  mainView: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderWidth: 1,
    borderColor: '#F4F4F4',
    borderRadius: 5,
    borderStyle: 'dashed',
    marginBottom: 5,
    width: '100%',
    padding: 16,
  },
  input: {
    width: '85%',
    backgroundColor: '#F4F4F4',
    borderRadius: 8,
    height: 40,
  },
  btn: {
    justifyContent: 'center',
    alignItems: 'center',
    width: 40,
    height: 40,
    borderRadius: 3,
    backgroundColor: '#67BFBB',
  },
  addTxt: {
    color: 'white',
  },
});
