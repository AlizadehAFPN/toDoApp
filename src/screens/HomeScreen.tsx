import * as React from 'react';
import {
  View,
  Text,
  FlatList,
  SafeAreaView,
  StyleSheet,
  Alert,
  Platform,
  Linking,
} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import * as LocalAuthentication from 'expo-local-authentication';
import {addTask, deleteTask, editTask} from '../redux/reducer/taskSlice';
import {changeBiometric} from '../redux/reducer/biometric';
import Modal from 'react-native-modal';
import InputItem from '../components/InputItem';
import ListItem from '../components/listItem';
import {startActivityAsync, ActivityAction} from 'expo-intent-launcher';

const HomeScreen = () => {
  const dispatch = useDispatch();
  const rxState = useSelector((toDoList: any) => toDoList);
  const [state, setState] = React.useState({
    taskTitle: '',
    editTitle: '',
    editId: null,
    isBiometricSupported: false,
    modalVisible: false,
  });

  React.useEffect(() => {
    // just to check if the device support biometric or not. I don't want to filter old device;
    (async () => {
      const compatible = await LocalAuthentication.hasHardwareAsync();
      setState(s => ({...s, isBiometricSupported: compatible}));
    })();
  }, []);
  console.log(state.isBiometricSupported, '----');
  const onAddPressed = async () => {
    // check if the title is not empty when adding
    if (state.taskTitle.length == 0) {
      Alert.alert('No title!', 'Please enter a title for your new task');
      return;
    }

    // ask for login or check if logged in before
    let biomet = await handleBiometric();
    if (biomet) {
      dispatch(
        addTask({
          id: new Date().valueOf(),
          name: state.taskTitle,
        }),
      );
      setState(s => ({...s, taskTitle: ''}));
    }
  };

  const onEditPressed = async () => {
    // check if the title is not empty
    if (state.editTitle.length == 0) {
      Alert.alert('No title!', 'Please enter a title for your edited task');
      return;
    }
    // ask for login or check if logged in before
    let biomet = await handleBiometric();

    if (biomet) {
      dispatch(
        editTask({
          id: state.editId,
          name: state.editTitle,
          editedtime: new Date().valueOf(),
        }),
      );
      setState(s => ({...s, modalVisible: false}));
    }
  };

  const onDeleteTask = async (item: any) => {
    // delete task from list by dispatching and updating redux store
    let biomet = await handleBiometric();
    if (biomet) dispatch(deleteTask(item));
  };

  const handleBiometric = async () => {
    // If the user logged in before or his/her device are not supporting biometrics , we let him procceed
    if (rxState?.biometric?.value || state.isBiometricSupported == false)
      return true;

    // Ask for Login with biometric
    try {
      const biometricAuth = await LocalAuthentication.authenticateAsync({
        promptMessage: 'Login with Biometrics',
        disableDeviceFallback: true,
        cancelLabel: 'Cancel',
      });
      //@ts-ignore
      if (biometricAuth?.error == 'not_available') {
        Alert.alert(
          'Attention!',
          'you haven not set a biometric yet, do you want to add it?',
          [
            {
              text: 'Cancel',
              onPress: () => console.log('Cancel Pressed'),
              style: 'cancel',
            },
            {
              text: 'OK',
              onPress: () => {
                if (Platform.OS == 'android')
                  startActivityAsync(ActivityAction.SECURITY_SETTINGS);
                else {
                  Linking.openURL('app-settings:');
                }
              },
            },
          ],
        );
      }
      // I just wanted the app work on simulator
      //@ts-ignore
      if (biometricAuth?.error == 'not_enrolled') {
        Alert.alert(
          'Warning!',
          'you are not using a secure device, Anyway, you can use our app :)',
          [
            {
              text: 'Cancel',
              onPress: () => console.log('Cancel Pressed'),
              style: 'cancel',
            },
            {
              text: 'OK',
              onPress: () => {
                bioSuccedd();
              },
            },
          ],
        );
      }
      //success login
      if (biometricAuth.success) {
        bioSuccedd();
        return true;
      }
    } catch (error) {
      console.log(error, '++++');
    }
  };

  const bioSuccedd = () => {
    // update redux for next steps
    dispatch(changeBiometric(true));
  };

  const onModalPressed = (item: any) => {
    // open modal with selected value
    setState(s => ({
      ...s,
      modalVisible: true,
      editTitle: item?.name,
      editId: item?.id,
    }));
  };

  const closeModal = () => {
    // just close modal
    setState(s => ({...s, modalVisible: false}));
  };

  const onChangeText = (taskTitle: string) => {
    setState(s => ({...s, taskTitle}));
  };

  const keyExtractor = React.useCallback(
    (item: any, index: number) => item._id || `${index}`,
    [],
  );

  return (
    <SafeAreaView
      style={{flex: 1, alignItems: 'center', backgroundColor: 'white'}}>
      <InputItem
        btnName={'Add'}
        value={state.taskTitle}
        onChangeText={onChangeText}
        onBTNPressed={onAddPressed}
        placeHolder={'Add new task'}
      />

      <FlatList
        data={rxState?.tasks}
        style={{width: '100%', alignSelf: 'center'}}
        contentContainerStyle={{width: '100%', alignSelf: 'center'}}
        keyExtractor={keyExtractor}
        renderItem={({item, index}) => (
          <ListItem
            index={index}
            length={rxState?.tasks.length}
            item={item}
            onModalPressed={onModalPressed}
            onDeleteTask={onDeleteTask}
          />
        )}
      />
      <Modal
        isVisible={state.modalVisible}
        backdropOpacity={0.5}
        onBackButtonPress={closeModal}
        onBackdropPress={closeModal}>
        <View style={{backgroundColor: 'white', borderRadius: 8}}>
          <Text style={{fontSize: 16, color: '#67BFBB', padding: 16}}>
            Edit Task :
          </Text>
          <InputItem
            value={state.editTitle}
            onChangeText={(name: string) =>
              setState(s => ({...s, editTitle: name}))
            }
            onBTNPressed={onEditPressed}
            placeHolder={'Edit task name'}
            btnName={'Edit'}
          />
        </View>
      </Modal>
    </SafeAreaView>
  );
};

export default HomeScreen;

const styles = StyleSheet.create({
  start: {},
  button: {},
});
