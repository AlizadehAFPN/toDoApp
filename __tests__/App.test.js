/**
 * @format
 */

import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import ListItem from '../src/components/listItem';
import InputItem from '../src/components/InputItem';
import * as TestRenderer from 'react-test-renderer';
import HomeScreen from '../src/screens/HomeScreen';

it('renders with no change', () => {
  renderer.create(<InputItem />);
});

it('renders with no change', () => {
  renderer.create(<ListItem />);
});


it('if we pass a value to the input, our placeholder should be empty', () => {
  const tree = renderer.create(<InputItem value={'test'}   />).toJSON();
  expect((tree?.children)[0]?.props.placeholder).toBe('')
});

it('if index more than 0, borderradius ahould be 0', () => {
  const list = renderer.create(<ListItem index={1}   />).toJSON();
  expect((list.props.style)[1]?.borderTopRightRadius).toBe(0)
});



