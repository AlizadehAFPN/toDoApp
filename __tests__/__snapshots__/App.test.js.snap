// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#EEEEEE",
        "borderColor": "#CACACA",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginHorizontal": 16,
        "padding": 16,
      },
      Object {
        "borderBottomLeftRadius": 0,
        "borderBottomRightRadius": 0,
        "borderTopLeftRadius": 0,
        "borderTopRightRadius": 0,
        "borderTopWidth": 1,
      },
    ]
  }
  testID="viewItem"
>
  <View
    style={
      Object {
        "width": "80%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
        }
      }
    >
      Name: 
    </Text>
    <Text
      style={
        Object {
          "marginTop": 8,
        }
      }
    >
      Created: 
      Invalid date
    </Text>
    <Text
      style={
        Object {
          "color": "#67BFBB",
          "marginTop": 8,
        }
      }
      testID="editedtest"
    >
      Last Update: 
      No Update Yet
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": "100%",
        "justifyContent": "flex-end",
        "width": "20%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 20,
            },
            Object {
              "marginRight": 16,
            },
            Object {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#CC0000",
              "fontSize": 20,
            },
            undefined,
            Object {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
